ihmc-communication/src/main/java/us/ihmc/communication/ros2/sync/ROS2PeerClockOffsetEstimator.java

```diff
@@ -0,0 +1,143 @@
+   private final SampleInfo sampleInfo = new SampleInfo();
+   private final Guid receivedRequestTarget = new Guid();
+   private final Guid receivedReplyTarget = new Guid();
+
+   public ROS2PeerClockOffsetEstimator(ROS2Node ros2Node)
+   {
+      publisher = ros2Node.createPublisher(TOPIC);
+      ourGuid = publisher.getPublisher().getGuid();
+
+      ros2Node.createSubscription(TOPIC, subscriber ->
+      {
+         subscriber.takeNextData(receivedMessage, sampleInfo);
+
+         MessageTools.fromMessage(receivedMessage.getRequestTarget(), receivedRequestTarget);
+         MessageTools.fromMessage(receivedMessage.getReplyTarget(), receivedReplyTarget);
+
+         if (receivedMessage.getIsRequest() && receivedRequestTarget.equals(ourGuid)) // Reply
+         {
+            PeerClockOffsetEstimatorPingMessage replyMessage = new PeerClockOffsetEstimatorPingMessage();
+            replyMessage.set(receivedMessage);
+            replyMessage.setIsRequest(false);
+            MessageTools.toMessage(Instant.now(), replyMessage.getReplySendTime());
+
+            ThreadTools.startAsDaemon(() ->
+            {
+               synchronized (publisher)
+               {
+                  publisher.publish(replyMessage);
+               }
+            }, getClass().getSimpleName() + "PublishReply");
+         }
+         else if (!receivedMessage.getIsRequest() && receivedReplyTarget.equals(ourGuid)) // Update clock offset estimate
+         {
+            ROS2PeerClockOffsetEstimatorPeer peer = peerMap.get(receivedRequestTarget);
+            if (peer != null)
+            {
+               peer.update(MessageTools.toInstant(receivedMessage.getRequestSendTime()),
+                           Instant.now(),
+                           MessageTools.toInstant(receivedMessage.getReplySendTime()));
+            }
+         }
+      }, (subscriber, info) ->
+      {
+         Guid guid = info.getGuid();
+         if (!guid.equals(publisher.getPublisher().getGuid())) // Exclude our publisher
+         {
+            switch (info.getStatus())
+            {
+               case MATCHED_MATCHING ->
+               {
+                  if (!peerMap.containsKey(guid))
+                  {
+                     Guid guidCopy = new Guid();
+                     guidCopy.set(guid);
+
+                     ROS2PeerClockOffsetEstimatorPeer peer = new ROS2PeerClockOffsetEstimatorPeer(guidCopy);
+                     peerMap.put(guidCopy, peer);
+                     peerList.add(peer);
+                  }
+               }
+               case REMOVED_MATCHING ->
+               {
+                  peerMap.remove(guid);
+                  peerList.removeIf(peer -> peer.getGuid().equals(guid));
+               }
+            }
+         }
+      });
+
+      requestThread.setFrequencyLimit(5.0);
+      requestThread.startRepeating();
+   }
+
+   private void requestThread()
```

Comment:
Can this be called requestUpdate or something? It doesn't implicitly have anything to do with a thread by itself

Reply:
changed to runRequestTask